src/daft-logical-plan/src/optimization/rules/extract_window_function.rs

```diff
@@ -0,0 +1,488 @@
+                                (e.clone(), semantic_id.id.to_string())
+                            })
+                            .collect::<Vec<(ExprRef, String)>>();
+
+                        window_col_mappings.extend(spec_mappings);
+                    }
+
+                    let new_projection = project
+                        .projection
+                        .iter()
+                        .map(|expr| Self::replace_window_functions(expr, &window_col_mappings))
+                        .collect::<DaftResult<Vec<ExprRef>>>()?;
+
+                    let final_plan = Arc::new(LogicalPlan::Project(Project::try_new(
+                        current_plan,
+                        new_projection,
+                    )?));
+
+                    Ok(Transformed::yes(final_plan))
+                } else {
+                    Ok(Transformed::no(node))
+                }
+            }
+            _ => Ok(Transformed::no(node)),
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use common_error::DaftResult;
+    use daft_dsl::{expr::window::WindowSpec, resolved_col, Expr};
+    use daft_schema::{dtype::DataType, field::Field};
+
+    use crate::{
+        logical_plan::{LogicalPlan, Project},
+        ops::Window,
+        optimization::{
+            optimizer::{RuleBatch, RuleExecutionStrategy},
+            rules::ExtractWindowFunction,
+            test::assert_optimized_plan_with_rules_eq,
+        },
+        test::{dummy_scan_node, dummy_scan_operator},
+    };
+
+    fn assert_optimized_plan_eq(
+        plan: Arc<LogicalPlan>,
+        expected: Arc<LogicalPlan>,
+    ) -> DaftResult<()> {
+        assert_optimized_plan_with_rules_eq(
+            plan,
+            expected,
+            vec![RuleBatch::new(
+                vec![Box::new(ExtractWindowFunction::new())],
+                RuleExecutionStrategy::Once,
+            )],
+        )
+    }
+
+    #[test]
+    fn test_single_partition_window_function() -> DaftResult<()> {
+        let scan_op = dummy_scan_operator(vec![
+            Field::new("category", DataType::Utf8),
+            Field::new("value", DataType::Int64),
+        ]);
+
+        let input_plan = dummy_scan_node(scan_op.clone());
+
+        let category_col = resolved_col("category");
+        let value_col = resolved_col("value");
+
+        let mut window_spec = WindowSpec::default();
+        window_spec.partition_by = vec![category_col.clone()];
+
+        let min_expr = value_col.clone().min();
+
+        let window_func =
+            Arc::new(Expr::Window(min_expr.clone(), window_spec.clone())).alias("min_value");
+
+        let projection = vec![category_col.clone(), value_col.clone(), window_func.clone()];
+
+        let plan = input_plan.clone().select(projection)?.build();
+
+        let auto_generated_name = "value.local_min().window(partition_by=[category],order_by=[])";
```

Comment:
you could just use the semantic ID here too, we probably dont want to test semantic ID logic here in case we ever change it